### ISSUE TITLE
[xml] missing <ptype> tags in <command>'s

### ISSUE THREAD

**Issue by SunSerega** [KhronosGroup/OpenGL-Registry#353]
In `gl.xml` only `void`, `void*` and `void**` return and parameter types are not marked with `<ptype>`.
This is inconvenient, but at least consistent.

But in other 2 files (`wgl.xml`, `glx.xml`) - `<ptype>` tag is missing for most (in not all) of the `int`, `float`, etc. parameters.

**Comment by SunSerega** [comment]
Unfortunately only after posting i realized i should only check .pdf:

> The <ptype> tag is optional, and contains text which is a valid type name found
in <type> tag, and indicates that this type must be previously defined for the
definition of the command to succeed. Builtin C types, and any derived types
which are expected to be found in other header files, should not be wrapped in
<ptype> tags

Still, `<ptype>` is usefull for creating wrappers in languages other then C++.

If for comparability you don't want to change usage of `<ptype>` - how about add another tag type? Like `<ntv_ptype>`, which would indicate that type is of core C++ lang, and not of gl itself.

**Comment by SunSerega** [comment]
Also, there is one function in glx.xml:
```
        <command>
            <proto><ptype>GLboolean</ptype> <name>glXSet3DfxModeMESA</name></proto>
            <param>GLint <name>mode</name></param>
        </command>
```
`GLint` is a gl type, so it supposed to have `<ptype>` tag, right?
And should i make separate issue for this?

**Comment by Perksey** [comment]
Having wrote roughly 4 parsers of the gl.xml for .NET-based languages, I can safely say that it is clearly not intended to work well with other languages. It is possible, just very hard. The XML schema likely won't change to reflect this as the schema is well-established with the current command type format, and other languages aren't really a concern for Khronos.

Side note: I believe the old schema actually did use ptypes for `<ptype>GLvoid</ptype>` but it was removed in favour of `void`, so I don't see the schema taking a step back to the old format any time soon. I agree it would be useful to downstream consumers, but it'd be too much of a breaking change.

> GLint is a gl type, so it supposed to have <ptype> tag, right?
And should i make separate issue for this?

Yes.

**Comment by SunSerega** [comment]
I think problem isn't that its unintended, but mostly that there is no one testing and reporting little details.

**Comment by Perksey** [comment]
Not having tags around `void` is definitely intended. All other types are intended to be `<ptype>` to my knowledge.

**Comment by SunSerega** [comment]
> I think problem isn't that its unintended

was directed to

> I can safely say that it is clearly not intended to work well with other languages

---

> Not having tags around void is definitely intended. All other types are intended to be <ptype> to my knowledge.

They are both currently intended because it's documented in ReadMe.pdf file, which can be found in same folder as .xml's. I already posted quote in second comment here.

What i'm asking is slightly changing .xml structure.
Adding new tag should break anything, because any good XML parser (like `System.Xml.XmlDocument` in .Net, which is what i'm currently using) is gonna remove tags from `.InnerText`-like property, which is what you would want to use to see content of `<proto>` or `<param>` tag if you are already parsing it yourself from C-like definition (which is also what i'm doing currently as a hotfix for this).

To have an example:
```
<param>int <name>x</name></param>
```
```
<param><ntv_ptype>int</ntv_ptype> <name>x</name></param>
```
`.InnerText` property of both of these tags would return same string:
```
int x
```
And while this doesn't break anything - it would really help creating definitions in other languages, because you don't need to parse anything (although all parsing in hotfix for this is 3 lines).

To emphasize why this would be really needed:
```
ret_t f1(par_t1 par1);
```
In pascal, which i'm making wrapper for, this would look like:
```
function f1(par1: par_t1): ret_t;
```
So everything is reversed. And to reverse words you need to first either parse (which what i already made before taking my time to find this repository and see that it's already mostly pre-parsed in form of xml's) or get it in pre-parsed state.

---

Also, C-like lang definitions don't need to have something like `<name>mode</name>`. If that would be only concern of the one making these xml-s - this just wouldn't be there. Because of this i think `<ptype>` and `<name>` tags were from the start intended to use in creating definitions in other languages.

**Comment by Perksey** [comment]
I don't think adding a new element type `ntv_ptype` nor readding the void ptype will fly with the working group as this reverts a decision they've already made.

If your proposal is that all non-void types should be marked with `ptype` then I 100% agree.

**Comment by SunSerega** [comment]
I personally don't care that much about `void`, if it's:
- The only type to not have a tag.
- And this behavior is reflected with clear language in pdf.

**Comment by SunSerega** [comment]
And i would be okay with any tag for non-void types, as long as it's also consistent. I proposed having separate `<ntv_ptype>` tag only because currently .pdf clearly sais that all types marked with `<ptype>` are also defined in same .xml file. And so theoretically using `<ptype>` for build-in types could break some code using said .xml's.

**Comment by Perksey** [comment]
Ah ok. All `ptypes` should only reference defined types in the `types` section at the top of the file, so if another type is referenced that's probably in error.

Please change the issue title to `Document void types not having ptypes` or something of that essence and then wait for comments from @oddhack.

**Comment by SunSerega** [comment]
It does look documented to me (though maybe not as clear as it could be). And there is just more to decide about (like making separate issue for non-void types, i was thinking about that too).

So i would prefer to first wait for response from someone actually responsible for this repository (which if i understand right - only regards @oddhack).

**Comment by oddhack** [comment]
These tags are intended to be used to satisfy dependency walking for the C header generation. They were not inserted with other languages in mind. So a GLboolean or GLint type gets tagged and the generator scripts follow the dependencies down to the point where a typedef for the underlying C language type is generated, like the

```<type>typedef int <name>GLint</name>;</type>```

definition near the top of gl.xml. Builtin C types don't actually have any dependency that needs to be emitted first, so they don't need to be tagged.

Someone generating bindings to non-C languages needs to actually parse the declarations as C and construct a representation they can more easily translate to Pascal or whatever the target is. The declarations are so limited that this could be done with regex matching - or if someone was sufficiently motivated, they could hook into compiler toolkit internals and get back an AST. Given that there are numerous bindings for other languages out there already I have to think this is a solved problem - but nobody has offered up their solution to share here, and that would probably be helpful for other folks, like the OP here. 

People have asked in the past if we would use an AST-in-XML representation instead of the C declarations, and the answer is no.

**Comment by SunSerega** [comment]
> These tags are intended to be used to satisfy dependency walking for the C header generation

You mean, to only put in .h types that are necessary for selected OpenGL version?

And what do you think about adding new tag, like `<ntv_ptype>` i described higher?

**Comment by Perksey** [comment]
> but nobody has offered up their solution to share here, and that would probably be helpful for other folks, like the OP here.

Aha allow me to step in. Feel free to use the [Silk.NET BuildTools](https://github.com/Ultz/Silk.NET/tree/master/src/Core/BuildTools) as a reference - it's MIT so don't worry about licenses. Not everything will be transferrable, as this targets C# using a very specific solution for dynamic loading, but you can take a look at it and see what you can salvage. Note that this hasn't been updated to use the new `group` attribute so take the enums stuff with a grain of salt.

> And what do you think about adding new tag, like <ntv_ptype> i described higher?

Of course we should still hold out for Jon's comment, but I don't imagine this will be favourable for Khronos as it's not their goal to make the gl.xml easy to use outside of Khronos purposes and/or C/C++ header generation. To solve the native parameter type problem, you're probably better off writing a typemap to use in your binder (like I've done [here](https://github.com/Ultz/Silk.NET/blob/master/build/binder_typemaps/gl.json) and [here](https://github.com/Ultz/Silk.NET/blob/master/build/binder_typemaps/csharp.json)) or, the more difficult option, manually stepping through all typedefs until you resolve a native type.

**Comment by oddhack** [comment]
> > These tags are intended to be used to satisfy dependency walking for the C header generation
> 
> You mean, to only put in .h types that are necessary for selected OpenGL version?

Mostly. The process of generating outputs is entirely dependency driven starting from a required set of versions and extensions, which in turn require commands and enums, and the command interfaces generally require GL data types.

> And what do you think about adding new tag, like `<ntv_ptype>` i described higher?

We're not likely to do that. Something like @Perksey's comment seems appropriate for your use case, and hopefully their code would be helpful, too. I think this issue has gone as far as it usefully can, so closing.
